feat(NgStyle): add new NgStyle directive

pkozlowski-opensource opened the pull request:
<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/angular/angular/2665)

<!-- Reviewable:end -->

pkozlowski-opensource committed:
feat(NgStyle): add new NgStyle directive

matanlurey commented on modules/angular2/src/directives/ng_style.ts:
```diff
@@ -0,0 +1,38 @@
+
+  set rawStyle(v) {
+    this._rawStyle = v;
+    this._pipe = this._pipeRegistry.get('keyValDiff', this._rawStyle);
```
I'm not familiar with the pipe concept yet, but does keyValDiff also support Arrays or only Maps?

(This was the Angular 1.0 behavior)

pkozlowski-opensource replied:
@matanlurey nope, `keyValDiff` works only with the Map-like objects. But reading your comment I'm getting an impression that you were thinking about `ngClass`, since I can see only maps being supported in ng1 (both [JS](https://docs.angularjs.org/api/ng/directive/ngStyle) and [Dart](https://docs.angulardart.org/#angular/angular-directive.NgStyle))

If you are interested in the `ngClass` equivalent, the relevant PR is this: #2664)

gkalpak replied:
FWIW, `ngStyle` in ng1 does support expressions (not only maps).

gkalpak replied:
(But only expressions that evaluate to maps, so it's not far off :smiley:)

pkozlowski-opensource replied:
Right, this is exactly what this version does as well :-)

matanlurey commented on modules/angular2/src/directives/ng_style.ts:
```diff
@@ -0,0 +1,38 @@
+
+  onCheck() {
+    var diff = this._pipe.transform(this._rawStyle);
+    if (isPresent(diff) && isPresent(diff.wrapped)) {
```
Do we have any benchmarks on how this could effect large Angular applications?

pkozlowski-opensource replied:
I don't think we've got any specific benchmarks for the directives discussed here, but my understanding is that the `onCheck` method should be called only when a _referenc_e to an object being watched changes - which should be pretty rare in case of `ngStyle`. The other part of the "speed" equation here would be performance of change detection for object properties - this I believe is already covered (?) in benchamrks. @vsavkin / @tbosch might know more.

vsavkin replied:
We don't have microbenchmarks testing transform or onCheck. But since this is how ngFor is implemented, pretty much every macrobenchmark tests it indirectly.

I don't expect the performance to be a problem here. The ".onCheck" call is monomorphic. The transform call will most likely be monomorphic too. 

>> but my understanding is that the onCheck method should be called only when a _referenc_e to an object being watched changes

onCheck is called every single time an object is checked. onCheck is used for a deep check when all the references stay the same.

matanlurey commented on modules/angular2/src/directives/ng_style.ts:
```diff
@@ -0,0 +1,38 @@
+    diff.forEachRemovedItem((record) => { this._removeStyle(record.key); });
+  }
+
+  private _setStyle(name, val) { DOM.setStyle(this._domEl, name, val); }
```
Can _setStyle and _removeStyle be typed (e.g. name: String, val: String)?

pkozlowski-opensource replied:
Yep, going to add types.

pkozlowski-opensource commented:
@naomiblack cleaned up code and addressed comments. This needs another round of review (I can see that @vsavkin is already assigned and he knows this part best)

vsavkin commented on gulpfile.js:
```diff
@@ -439,7 +439,7 @@ gulp.task('!test.unit.js/karma-run', function(done) {
   runSequence(
     'build/tree.dart',
-    '!build/pubget.angular2.dart',
+    //'!build/pubget.angular2.dart',
```
Should it be commented out?

pkozlowski-opensource replied:
Nope, good catch!

vsavkin commented on modules/angular2/src/directives/ng_style.ts:
```diff
@@ -0,0 +1,36 @@
+
+@Directive({selector: '[ng-style]', lifecycle: [onCheck], properties: ['rawStyle: ng-style']})
+export class NgStyle {
+  _pipe;
```
types
